test/e2e/storage/testsuites/volumemode.go

```diff
@@ -293,7 +333,48 @@ func (t *volumeModeTestSuite) defineTests(driver TestDriver, pattern testpattern
 				ginkgo.By("Checking if read/write to persistent volume works properly")
 				utils.CheckReadWriteToPath(pod, pattern.VolMode, "/mnt/volume1")
 			})
-			// TODO(mkimuram): Add more tests
+
+			ginkgo.It("should fail to use a volume in a pod with mismatched mode [Slow]", func() {
+				skipBlockTest(driver)
+				init()
+				defer cleanup()
+
+				var err error
+
+				ginkgo.By("Creating sc")
+				l.sc, err = l.cs.StorageV1().StorageClasses().Create(l.sc)
+				framework.ExpectNoError(err)
+
+				ginkgo.By("Creating pv and pvc")
+				l.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Create(l.pvc)
+				framework.ExpectNoError(err)
+
+				err = framework.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, l.cs, l.pvc.Namespace, l.pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
+				framework.ExpectNoError(err)
+
+				l.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.pvc.Namespace).Get(l.pvc.Name, metav1.GetOptions{})
+				framework.ExpectNoError(err)
+
+				l.pv, err = l.cs.CoreV1().PersistentVolumes().Get(l.pvc.Spec.VolumeName, metav1.GetOptions{})
+				framework.ExpectNoError(err)
+
+				ginkgo.By("Creating pod")
+				pod := framework.MakeSecPod(l.ns.Name, []*v1.PersistentVolumeClaim{l.pvc}, false, "", false, false, framework.SELinuxLabel, nil)
+				// Change volumeMounts to volumeDevices and the other way around
+				pod = swapVolumeMode(pod)
+
+				// Run the pod
+				pod, err = l.cs.CoreV1().Pods(l.ns.Name).Create(pod)
+				framework.ExpectNoError(err)
+				defer func() {
+					framework.ExpectNoError(framework.DeletePodWithWait(f, l.cs, pod))
+				}()
+
+				// TODO: find a faster way how to check the pod can't start,
+				// perhaps when https://github.com/kubernetes/kubernetes/issues/79794 is fixed.
```

Comment:
I think that we decided not to rely on events in e2e tests [here](https://github.com/kubernetes/kubernetes/issues/71570), because it makes tests flakey.
Although, I agree that we need to find a good way to detect expected failure in the future.

Reply:
I agree that events may be unreliable, Failed pod status would be better. Let's fix that in #79794.

Reply:
With #79794, events are the only option. I added their processing to the test, but made sure that a missed event won't harm the test in any way. Events only speed up the test, we still check the pod is not scheduled if all events are lost.